### PR TITLE
Jetpack Visibility i1: Show jetpack scan history for atomic site only

### PR DESCRIPTION
**Pull request by vykes-mac**
Closes [#7268](https://github.com/Automattic/dotcom-forge/issues/7268)

## Proposed Changes

* Hide navigation on scan history page
* Allow browsing of history page only on atomic sites

image

## Why are these changes being made?

* Highlighting the benefits of Jetpack Scan which is already included in users' plans pet6gk-163-p2

## Testing Instructions

* On atomic site browse `/scan/history/:siteSlug`
* Verify scan history is shown
* Verify scan history is not shown for simple sites

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

File: client/my-sites/scan/controller.js [hunk 1]
```diff
@@ -97,8 +98,11 @@ export function scan( context, next ) {
 }
 
 export function scanHistory( context, next ) {
+	const state = context.store.getState();
+	const siteId = getSelectedSiteId( state );
+	const showNavigation = ! isAtomicSite( state, siteId );
```
Comment (p-jackson) [inline client/my-sites/scan/controller.js]: Just a thought that we don't want to accidentally start hiding the navigation tabs on `cloud.jetpack.com`. From my testing I couldn't make the **Scan** menu item appear for a WoA site, so maybe it's not possible. But if it ever did become possible to have a WoA site on Jetpack Cloud with the Scan menu item enabled, we wouldn't want to make unintended changes.

```suggestion
	const showNavigation = isJetpackCloud() || ! isAtomicSite( state, siteId );
```

File: client/my-sites/scan/index.js [hunk 2]
```diff
@@ -16,16 +16,17 @@ import {
...
 const notFoundIfNotEnabled = ( context, next ) => {
 	const state = context.store.getState();
 	const siteId = getSelectedSiteId( state );
 	const showJetpackSection = isJetpackSectionEnabledForSite( state, siteId );
-	const isWPCOMSite = getIsSiteWPCOM( state, siteId );
+	const isHistoryScanForAtomicSite =
+		isAtomicSite( state, siteId ) && context.path.includes( '/scan/history' );
 
-	if ( isWPCOMSite || ( ! isJetpackCloud() && ! showJetpackSection ) ) {
+	if ( ! isHistoryScanForAtomicSite || ( ! isJetpackCloud() && ! showJetpackSection ) ) {
```
Comment (p-jackson) [inline client/my-sites/scan/index.js]: This logic is starting to get a bit tricky for me to wrap my head around. If we have logic that should only run on the `/scan/history` route, then should that be handled by having a different controller?

The behaviour we want for atomic sites seems different enough that the `notFoundIfNotEnabled` name no longer seems to fit.

Would it be cleaner to define a whole different controller for the scan history case?

```ts
const notFoundIfNotAtomic = ( context, next ) => {
	const state = context.store.getState();
	const siteId = getSelectedSiteId( state );
	const showJetpackSection = isJetpackSectionEnabledForSite( state, siteId );

	if ( ! isAtomicSite( state, siteId ) || ( ! isJetpackCloud() && ! showJetpackSection ) ) {
		return notFound( context, next );
	}

	next();
};
```

Maybe? Maybe not, what do you think? It's a pity about duplicating the logic for "isJetpackSectionEnabled", but I find the logic a little easier to follow (not much though).

Comment (p-jackson) [reply client/my-sites/scan/index.js]: What about a little function currying 🌶️ 

```ts
const notFoundIfNotEnabled =
	( { allowOnAtomic } ) =>
	( context, next ) => {
		const state = context.store.getState();
		const siteId = getSelectedSiteId( state );
		const showJetpackSection = isJetpackSectionEnabledForSite( state, siteId );
		const disabled = allowOnAtomic
			? getIsSiteWPCOM( state, siteId ) && ! isAtomicSite( state, siteId )
			: getIsSiteWPCOM( state, siteId );

		if ( disabled || ( ! isJetpackCloud() && ! showJetpackSection ) ) {
			return notFound( context, next );
		}

		next();
	};

// ...

	page(
		'/scan/history/:site/:filter?',
		siteSelection,
		// ...
		notFoundIfNotEnabled( { allowOnAtomic: true } ),
		// ...
	);
```

**Review by p-jackson — COMMENTED** [review]
Change mostly works for me, I just find the code a little hard to follow; see below.

I did find one odd case during testing. If the site is atomic, but it doesn't have the correct plan, then we'll show this upsell. 

CleanShot 2024-05-21 at 20 42 52@2x

This happened for me because I had left my store sandboxed. But the same thing would happen when a site's Creator plan has expired but we haven't reverted the site to Simple yet. Maybe `/scan/history` should be disabled if the site's plan doesn't support the scan feature?

File: client/my-sites/scan/controller.js (hunk 1 above)
Comment (vykes-mac) [reply client/my-sites/scan/controller.js]: Yeah it make sense to safeguard. The assumption was that `isJetpackCloud()` and `isAtomic` are always mutually exclusive

File: client/my-sites/scan/index.js (hunk 2 above)
Comment (vykes-mac) [reply client/my-sites/scan/index.js]: > What about a little function currying 🌶️

Yeah I find this more accommodating that than creating a whole new controller for scan history routes.

**Comment by vykes-mac** [comment]
Thanks @p-jackson for the suggested improvements, I have applied those changes.

In regards to 

> This happened for me because I had left my store sandboxed. But the same thing would happen when a site's Creator plan has expired but we haven't reverted the site to Simple yet. Maybe /scan/history should be disabled if the site's plan doesn't support the scan feature?

It seems to only appear when the store is sandboxed. I tested it with a site that has an expired plan before downgrading to atomic and it still shows the scan history which I think make sense since the atomic data would be available still until revert.

**Review by p-jackson — APPROVED** [review]
👍

> It seems to only appear when the store is sandboxed.

Thanks for checking
